Add support for ECDSA keys in crypto.pkcs11.Key

mbrossard opened the pull request:
This submission adds support for ECDSA (P224, P256, P384, P521) for crypto.pkcs11.Key. This code was tested with a smart-card (P256) and a software token (other curves).

kisom commented on crypto/pkcs11key/key.go:
```diff
@@ -223,39 +282,133 @@ func (ps *Key) getPrivateKey(module ctx, session pkcs11.SessionHandle, label str
-			bigE.SetBytes(a.Value)
-			e = int(bigE.Int64())
-			gotExponent = true
+func (ps *Key) getPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkcs11.ObjectHandle) (interface{}, error) {
```
This is now a huge function. It might be better to split it into getRSAPublicKey and getECPublicKey.

mbrossard replied:
Ok I can do that

kisom commented:
@jsha Would you mind looking this over as well?

kisom commented on crypto/pkcs11key/key.go:
```diff
@@ -198,10 +241,26 @@ func (ps *Key) getPrivateKey(module ctx, session pkcs11.SessionHandle, label str
+		if i == 0 && len(attribute.Value) > 0 {
+			if attribute.Value[0] == 0x3 {
+				ps.keyType = EC
+			} else {
```
Should probably explicitly check for RSA and error if it doesn't match.

jsha commented on crypto/pkcs11key/key.go:
```diff
@@ -198,10 +241,28 @@ func (ps *Key) getPrivateKey(module ctx, session pkcs11.SessionHandle, label str
+	if err != nil {
+		return noHandle, err
+	}
+	for i, attribute := range attributes {
```
Why loop over all attributes if you're only going to look at the first one?

mbrossard replied:
I tried to stay consistent with the rest of the file. The Block reading CKA_ALWAYS_AUTHENTICATE did it like that.

jsha commented on crypto/pkcs11key/key.go:
```diff
@@ -222,10 +283,9 @@ func (ps *Key) getPrivateKey(module ctx, session pkcs11.SessionHandle, label str
-func getPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkcs11.ObjectHandle) (rsa.PublicKey, error) {
-	var noKey rsa.PublicKey
+// Get the public key matching an RSA private key
+func (ps *Key) getRSAPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkcs11.ObjectHandle) (interface{}, error) {
```
You made this function a method on `Key`, but it doesn't actually access any members of `Key`. My understanding of Go idiom is that functions are preferred in such cases.

jsha replied:
Also, I think the return type shouldn't be `interface{}`.

mbrossard replied:
True for the methods, I'm converting them to functions. For the return type this relates to the next question.

jsha commented on crypto/pkcs11key/key.go:
```diff
@@ -76,11 +111,15 @@ type Key struct {
 
 	// The public key corresponding to the private key.
-	publicKey rsa.PublicKey
+	rsaPublicKey rsa.PublicKey
```
Why not change the type to `crypto.PublicKey`? That way you can avoid having two separate fields, plus the `keyType`. When you need to know the key type, e.g. in `Sign()` you can do a switch statement over the type.

mbrossard replied:
I've tried this and couldn't get it to work. Once you assign it to a crypto.PublicKey, crypto.x509.signingParamsForPublicKey() fails.

The problem also appears when changing the return types of the getPublicKey functions.

jsha replied:
I've run into errors like this a few times. As I understand it, the problem is that `crypto.PublicKey` is satisfied by both `rsa.PublicKey` and `*rsa.PublicKey`, but the code in `x509.go` that switches on types only understands the pointer types. In other words, any time you want to pass in a `crypto.PublicKey` to x509, you should plan on making the "real" type a pointer type.

I was able to make this work locally by removing an `&` in `Public()` and adding one in `setup`. Diff:

```
@@ -76,7 +77,7 @@ type Key struct {
 	pin string
 
 	// The public key corresponding to the private key.
-	publicKey rsa.PublicKey
+	publicKey crypto.PublicKey
 
 	// The an ObjectHandle pointing to the private key on the HSM.
 	privateKeyHandle pkcs11.ObjectHandle
@@ -171,7 +172,7 @@ func (ps *Key) setup(privateKeyLabel string) (err error) {
 		ps.module.CloseSession(session)
 		return
 	}
-	ps.publicKey = publicKey
+	ps.publicKey = &publicKey
 
 	return
 }
@@ -318,7 +319,7 @@ func (ps *Key) openSession() (pkcs11.SessionHandle, error) {
 
 // Public returns the public key for the PKCS #11 key.
 func (ps *Key) Public() crypto.PublicKey {
-	return &ps.publicKey
+	return ps.publicKey
 }
 
 // Sign performs a signature using the PKCS #11 key.
```

mbrossard replied:
Thanks for the tip, I've been banging my head against the wall on this one. I'll commit the change soon.
If it wasn't for the email notification, this comment would have been buried in the thread.

jsha commented on crypto/pkcs11key/key.go:
```diff
@@ -252,10 +312,117 @@ func getPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkc
+	id := []byte {}
+	gotOid, gotID := false, false
+	for _, a := range attr {
+		if a.Type == pkcs11.CKA_EC_PARAMS {
```
Please link to documentation on the format of `CKA_EC_PARAMS`.

jsha replied:
Also, is CKA_EC_PARAMS generally present on both the private and public key, or just the private one?

mbrossard replied:
CKA_EC_PARAMS in our case would be the DER encoded OID of the Curve, the namedCurve option, see [ECDSA public key objects](http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/os/pkcs11-curr-v2.40-os.html#_Toc416960012).

jsha commented on crypto/pkcs11key/key.go:
```diff
@@ -252,10 +312,117 @@ func getPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkc
+		if a.Type == pkcs11.CKA_EC_PARAMS {
+			oid = a.Value
+			gotOid = true
+		} else if a.Type == pkcs11.CKA_ID {
```
The code that searches for RSA keys doesn't pull the CKA_ID attribute. Will it always be present? Is it advisable to use? If so, should it be added to the RSA code?

mbrossard replied:
The RSA private key already has all the public key information, so there's no reason to search for the corresponding PKCS#11 public key object.
There's no guarantee that CKA_ID will be present but it's the most common way to link public and private keys.

jsha commented on crypto/pkcs11key/key.go:
```diff
@@ -252,10 +312,117 @@ func getPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkc
+		}
+	}
+	if !gotOid || !gotID {
+		return noKey, errors.New("public key missing either curve parameters or id")
```
Please split this into separate error cases with more specific messages.

mbrossard replied:
Ok (I'll do that also for RSA from which I copied the style)

jsha commented on crypto/pkcs11key/key.go:
```diff
@@ -252,10 +312,117 @@ func getPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkc
+	if err := module.FindObjectsInit(session, templateSearch); err != nil {
+		return noKey, err
+	}
+	objs, _, err := module.FindObjects(session, 2)
```
This is something I meant to look into / fix in the RSA code: the second parameter of `FindObjects` is `max`, the max number of objects to be returned. I think it should be one, since we only care about one object. Would you mind fixing that here and in the RSA branch?

mbrossard replied:
I agree with you assessment. I'll add that.

jsha commented on crypto/pkcs11key/key.go:
```diff
@@ -252,10 +312,117 @@ func getPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkc
+		return noKey, err
+	}
+
+	data := []byte {}
```
I believe Go style is `[]byte{}`. You can run `gofmt -w key.go` to fix any style issues automatically.

mbrossard replied:
I'll run `gofmt ` but this specific code is disappearing.

jsha commented on crypto/pkcs11key/key.go:
```diff
@@ -252,10 +312,117 @@ func getPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkc
+	}
+
+	x, y := elliptic.Unmarshal(curve, data)
+	if x == nil && y == nil {
```
Per docs for `elliptic.Unmarshal`: "On error, x = nil". It's confusing to check for both; better to just check for the one that indicates the error condition.

Can you also document the fallback parsing that is done below? When do you expect it to trigger, and what format does it expect? Would be good to break out this part of the function (from line 399 to line 407) into its own function for easier testing.

mbrossard replied:
Ok. PKCS#11 v2.20 specified the EC Point was supposed to be store in an ASN.1 OCTET STRING. I'll break it out and simplify.

jsha commented on crypto/pkcs11key/key.go:
```diff
@@ -307,8 +474,12 @@ func (ps *Key) openSession() (pkcs11.SessionHandle, error) {
 		if err = ps.module.Login(session, pkcs11.CKU_USER, ps.pin); err != nil {
-			ps.module.CloseSession(session)
-			return session, err
+			if err == pkcs11.Error(pkcs11.CKR_USER_ALREADY_LOGGED_IN) {
```
Can you add a comment describing when this condition would be encountered?

mbrossard replied:
This relates to the comment above. I just happen to use a token that doesn't support multiple login to the same session. I'll add a comment.

jsha replied:
I have a token like that as well: a Yubikey Neo in PIV mode. As the code is currently written, I think it's not safe to instantiate multiple `Key`'s. After each Sign, `Key` will call `Logout`. That means that if you instantiate two `Key`s, one will `Login` sucessfully, and with this code, the second will silently succeed even though the login attempt failed. However, after the first `Key` signs something, the second `Key` will think it's logged in but not be.

Generally I work around this by passing `-cpus 1` to pkcs11key_bench_test when I run it on my Yubikey.

So, my recommendation would be to pull this code out for this change. I'm definitely open to being convinced this can work, but I also think it's best as a separate branch.

mbrossard replied:
I don't think my change would break anything that wouldn't break before. The example you cite will fail with and without my change: once a call is made to `C_Logout` all other sessions on the token will fail.

In http://docs.oasis-open.org/pkcs11/pkcs11-base/v2.40/pkcs11-base-v2.40.html
"Since all sessions an application has with a token have a shared login state, `C_Login` only needs to be called for one of the sessions". The section about `C_Logout` is not very readable, but it does state that all sessions of the application are affected and all object handles to private objects are destroyed.

I've found that [OpenSSH](https://github.com/openssh/openssh-portable/blob/master/ssh-pkcs11.c#L380) and [GnuTLS](https://gitlab.com/gnutls/gnutls/blob/master/lib/pkcs11.c#L2552) treat `CKR_USER_ALREADY_LOGGED_IN` as equivalent and CKR_OK for `C_Login`.

I think this only highlights the fact that the login state should be tracked by token (here with multiple sessions for the same key, but also in the case of multiple keys on the same token). A solution that is clearly beyond this PR. Do you want me to open separate issue for login state management?

I'll move this patch to a different PR.

jsha replied:
Yes, definitely agree we should probably re-work how we deal with login
state. And thanks for moving this to another branch.

jsha commented on crypto/pkcs11key/key.go:
```diff
@@ -357,14 +531,21 @@ func (ps *Key) Sign(rand io.Reader, msg []byte, opts crypto.SignerOpts) (signatu
+
+	if ps.keyType == RSA {
+		mechanism = []*pkcs11.Mechanism{pkcs11.NewMechanism(pkcs11.CKM_RSA_PKCS, nil)}
+		prefix, ok := hashPrefixes[hash]
```
Why doesn't the hash prefix apply to EC signatures?

mbrossard replied:
The hashPrefixes is a table of pre-encoded ASN.1 prefix for DigestInfo objects used in RSA PKCS#1 1.5 signature (see src/pkg/crypto/rsa/pkcs1v15.go). CKM_ECDSA operates on raw hash values.

jsha commented on crypto/pkcs11key/key.go:
```diff
@@ -252,10 +312,117 @@ func getPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkc
+}
+
+// Get the public key matching an Elliptic Curve private key
+func (ps *Key) getECPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkcs11.ObjectHandle) (interface{}, error) {
```
This function is a bit on the long side and would probably benefit from being split into two, perhaps around `publicKeyHandle := objs[0]`.

jsha commented:
Thank you for working on this. This generally looks great, and I'm very excited about it because I want to support EC signatures for Let's Encrypt in 2016, so we'll need this code.

Could you please add tests for the new code?

mbrossard commented:
One of the reason I've looked at this is specifically because I saw Let's Encrypt wanted to support ECDSA Certification Authorities, so I looked at `boulder` and `cfssl` and saw that it didn't work for PKCS#11 case.

I wanted to add some tests (I did most of my tests using pkcs11key_bench_test.go which can work with either RSA or EC keys), but I didn't find any automated test done on PKCS#11 features. I was expecting to find SoftHSM or equivalent. I don't have enough experience on `cfssl` or golang to organize such a task, but I could contribute.

As an alternative to SoftHSM, I've built a small wrapper to make the NSS Soft Token easy to use (it has essentially one quirk). You can find it in https://github.com/mbrossard/pkcs11 (along with enough tools to initialise the token and generate a key).

jsha commented on crypto/pkcs11key/key.go:
```diff
@@ -198,10 +241,28 @@ func (ps *Key) getPrivateKey(module ctx, session pkcs11.SessionHandle, label str
+	if err != nil {
+		return noHandle, err
+	}
+	for _, attribute := range attributes {
```
This version of the loop will now examine each attribute, and error out if any attribute is other than CKK_RSA or CKK_EC. I was thinking more along the lines of: check that `len(attributes) > 0`, then look just at `attributes[0]`. I think in the common case where there's just one attribute matching CKA_KEY_TYPE it's the same, but that makes the code a little clearer.

mbrossard replied:
Adding that

jsha commented:
Regarding tests: yes, SoftHSM tests would be terrific, and I've been meaning to implement end-to-end(ish) tests like that for Boulder. However, in this case I was asking more about unittests. You can see an example in key_test.go.

Mocking out libraries in Go is pretty unintuitive if you are coming from another language. Generally you define an interface that represents the subset of upstream methods you depend on. In this case, that's `type ctx interface` in key.go. `pkcs11.Ctx` objects satisfy the interface, and that's what we use when running the code. However, in `key_test.go` we can define other types that also satisfy the interface, and inject them into `Key` objects for testing.

`mockCtx` is one of these types, and so is `mockCtxFailsAlaysAuthenticate`. For many test cases you'll probably want to define a new mock type that extends (in Go terms, 'embeds') `mockCtx` and adds the behaviors you want for that specific test.

```
// This is a version of the mock that gives CKR_ATTRIBUTE_TYPE_INVALID when
// asked about the CKA_ALWAYS_AUTHENTICATE attribute.
type mockCtxFailsAlwaysAuthenticate struct {
	mockCtx
}
```

It takes a little while to get up to speed on Go testing. If you want to do any live chat, I'm [@j4cob](https://twitter.com/j4cob) on Twitter and my DMs are open.

jsha commented on crypto/pkcs11key/key.go:
```diff
@@ -198,10 +241,28 @@ func (ps *Key) getPrivateKey(module ctx, session pkcs11.SessionHandle, label str
+			} else if attribute.Value[0] == pkcs11.CKK_EC {
+				ps.keyType = EC
+			} else {
+				return noHandle, fmt.Errorf("Unsupported key type %d", attribute.Value)
```
Shouldn't this be `attribute.Value[0]`?

mbrossard replied:
Changing that

jsha commented:
Is this ready for re-review?

mbrossard commented:
I think I responded to all the comments you made. I'll let you judge if they were properly addressed.

jsha commented on crypto/pkcs11key/key.go:
```diff
@@ -166,12 +204,18 @@ func (ps *Key) setup(privateKeyLabel string) (err error) {
 	}
-	ps.publicKey = publicKey
+	if ps.keyType == EC {
+		t := publicKey.(ecdsa.PublicKey)
```
I think you can delete this chunk of code if you change the return statements in `get{EC,RSA}PublicKey` to return a pointer.

I believe that allows you to drop the `keyType` member entirely. I'd recommend moving lines 245-264, in `getPrivateKey`, where you retrieve the key type, into their own function and calling that function from `getPublicKey`. Then `keyType` becomes a purely local variable, used only decide which of `getRSAPublicKey`, `getECPublicKey` to call.

mbrossard replied:
We still need the keyType for to choose the right signature mechanism.

jsha replied:
Good point. In theory we could do that with a switch statement like `switch pub := pub.(type) {` switch statement, but I don't feel strongly about that.

I do still think the change to emit pointers rather than values directly from `getPublicKey` is worthwhile, along with putting the `keyType` decision inside `getPublicKey`. What about this as a method signature:

`publicKey, keyType, err := ps.getPublicKey(...)`

mbrossard replied:
I'll try to change to pointers and see if the switch on the publicKey works too.

jsha commented:
Thanks for adding the tests! Please add tests for these specific cases:

- readECPoint produces correct output for default case
- readECPoint produces correct output for x == nil case
- signatureInput is appropriate for key type (has prefix for RSA ; has no prefix for ECDSA)
- error when no public key matching an EC key's CKA_ID is found
- after constructing an RSA key: rsaKeyPtr, ok := ps.Public().(*rsa.PublicKey); ok
- after constructing an EC key: ecKeyPtr, ok := ps.Public().(*ecdsa.PublicKey); ok
- X, Y, and Curve fields of constructed EC key are correct.

Sorry to be so demanding on the tests front, especially since the code as it stands doesn't have many tests. We very much want to improve that overall for Boulder, and for pkcs11key.

Note: checking for correct signatureInput is a little tricky. It probably requires having the mock save the latest bytes it was asked to sign as a member variable; or alternately, make the mock's sign method simply copy it's input as its output.

mbrossard commented:
I removed the `keyType` pushed the CKA_KEY_TYPE loading in the `loadPublicKey` method which replaces `getPublicKey`. That's the cleanest way I found in order to not have to due an extra switch/case on the key type value.

I also added all the tests you listed (except the part of one that I had to remove because it doesn't work on Go 1.4). I used the method of returning the input of Sign as the signature.

This should cover your latest comments.

jsha commented on crypto/pkcs11key/key.go:
```diff
@@ -249,13 +268,173 @@ func getPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkc
-	if !gotModulus || !gotExponent {
-		return noKey, errors.New("public key missing either modulus or exponent")
+	if !gotModulus {
+		return noKey, errors.New("private key missing modulus")
```
*public key missing modulus. Also below.

mbrossard replied:
We actually look in the private key so the message might be confusing. I'm suggesting: key is missing public {modulus,exponent}.

jsha commented on crypto/pkcs11key/key.go:
```diff
@@ -249,13 +268,173 @@ func getPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkc
 		E: e,
-	}, nil
+	}
+	return rsa, nil
```
You should return &rsa here. That way you won't have to do the cast, below.

mbrossard replied:
I tried earlier but failed to make it work. This time I go it to work. Nice!

jsha commented on crypto/pkcs11key/key.go:
```diff
@@ -249,13 +268,173 @@ func getPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkc
+		if err != nil {
+			return err
+		}
+		rsaPub := pub.(rsa.PublicKey)
```
This is the case you can avoid by returning `&rsa` above. Same thing applies for EC.

mbrossard replied:
This will be in the next push.

jsha commented on crypto/pkcs11key/key.go:
```diff
@@ -249,13 +268,173 @@ func getPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkc
+		return err
+	}
+
+	switch keyType {
```
I believe a simpler version of this would be (since err is already declared):

```
switch keyType {
  case pkcs11.CKK_RSA:
    ps.publicKey, err = getRSAPublicKey(...)
  case pkcs11.CKK_EC:
    ps.publicKey, err = getECPublicKey(...)
  default:
    err = fmt.Errorf(...)
}
return err
```

mbrossard replied:
This will be in the next push.

jsha commented on crypto/pkcs11key/key_test.go:
```diff
@@ -5,21 +5,32 @@ package pkcs11key
 var tokenInfo = pkcs11.TokenInfo{
 	Label: "token label",
 }
+var currentSearch []*pkcs11.Attribute
```
This should be an attribute of `mockCtx` rather than a global variable. That way multiple tests can run in parallel.

mbrossard replied:
This will be in the next push.

jsha commented on crypto/pkcs11key/key_test.go:
```diff
@@ -30,14 +41,39 @@ func (c mockCtx) FindObjectsFinal(sh pkcs11.SessionHandle) error {
+	for _, a := range currentSearch {
+		// We search private keys using CKA_LABEL
+		if a.Type == pkcs11.CKA_LABEL {
+			if a.Value[0] == 0x72 { // Label start with 'r'
```
I think this would be clearer as:

```
switch string(a.Value[0]) {
  case "rsa":
    return []pkcs11.Objecthandler{rsaPrivateKeyHandle}, true
  ...
```

mbrossard replied:
This will be in the next push. I also changed other sequences of if/else if for switch/case. With one small utility function I reduced the several functions shorter. I removed several magic values, replacing them with more explicit constants.

jsha commented on crypto/pkcs11key/key_test.go:
```diff
@@ -84,47 +257,96 @@ func (c mockCtx) SignInit(sh pkcs11.SessionHandle, m []*pkcs11.Mechanism, o pkcs
+	// Trying to load private EC key with no public key
+	err := k.setup("no_pub_ec")
+	if err == nil {
+		t.Fatalf("Unexpected succes")
```
*success. Also below.

This should probably be an Errorf instead of a Fatalf, because the next test case can still run and produce meaningful output even if this one fails. That allows a single test run to produce more info.

jsha replied:
(FWIW, I realize I didn't do the Errorf/Fatalf distinction properly in my previous testing code. Mea culpa :-) )

mbrossard replied:
This will be in the next push.

jsha commented on crypto/pkcs11key/key_test.go:
```diff
@@ -84,47 +257,96 @@ func (c mockCtx) SignInit(sh pkcs11.SessionHandle, m []*pkcs11.Mechanism, o pkcs
+
+	pub := ps.Public()
+	// Check public key is of right type
+	_ = pub.(*rsa.PublicKey)
```
```
_, ok := pub.(*rsa.PublicKey)
if !ok {
  t.Errorf("Attempted to load RSA key from module, got key of type %s. Expected *rsa.PublicKey", reflect.TypeOf(pub))
}
```

mbrossard replied:
This will be in the next push.

jsha commented:
Generally looking quite nice, thanks for all the work on this!

mbrossard commented:
I think we're really close to the end (if not there already).

Thank you for the help.

jsha commented on crypto/pkcs11key/key_test.go:
```diff
@@ -46,85 +99,219 @@ func (c mockCtx) GetAttributeValue(sh pkcs11.SessionHandle, o pkcs11.ObjectHandl
+		t.Errorf("Attempted to load ECDSA key from module, got key of type %s. Expected *ecdsa.PublicKey", reflect.TypeOf(pub))
+	}
+	if ecPub == nil ||
+		!(bytes.Equal(ecPub.X.Bytes(), ecPoint[3:35]) &&
```
This part, testing decoding of an EC point, would be better as its own `func TestReadECPoint` that calls `readECPoint` directly on a `[]byte`. Also, there are two cases in `readECPoint`: one where `elliptic.Unmarshal(curve, ecpoint)` succeeds, and one where there has to be an `asn1.Unmarshal` first. I assume this is testing the former. Could you add an additional `func TestReadECPointWithASNUnmarshal` to test the latter?

mbrossard replied:
The EC point in test has the ASN.1 wrapper already so that it tests the most code. I'm adding ASN.1 to the invalid key as well. I added some `readECPoint` tests.

jsha commented on crypto/pkcs11key/key_test.go:
```diff
@@ -46,85 +99,219 @@ func (c mockCtx) GetAttributeValue(sh pkcs11.SessionHandle, o pkcs11.ObjectHandl
+	}
+
+	// Trying to load private EC key with no public key
+	err := k.setup("no_publick_key_ec")
```
*no_public_key_ec.

This passed because we generated an error, but it (presumably) wasn't the correct error. I'm reluctant to suggest doing string matches on error messages, but it might be a good idea in this case.

mbrossard replied:
I'm adding this to the two tests with expected errors.

jsha commented:
Alright, just a last couple of comments on the test. This is looking great.

jsha commented:
Alright, looks good to me, thanks! @kisom, can you merge?

kisom commented:
LGTM. Thanks @mbrossard for this PR, and thanks @jsha for the thorough review!

mbrossard commented:
Thank you @jsha and @kisom. I'll look to see there's anything else I can contribute with.
